### DISCUSSION TITLE
"422 Repository creation failed" error with flux cli v0.12.x

### DISCUSSION THREAD

**Discussion by MxNxPx** [fluxcd/flux2#1302] [Q&A] [answered]
hi all,

i have been experimenting with flux2 using kind clusters and noticed that using newer versions of the flux cli -- namely newer than v0.11.0 -- i am no longer able to bootstrap a cluster using an existing github repo (if i revert the cli to v0.11.0 it works).

is there a flag to tell the flux cli on bootstrap to not try and create a new github repo but use the one that exists (and supplied)?
https://toolkit.fluxcd.io/guides/installation/#__code_6

here is what i am running and error seen...

```
$ kubectl config current-context
kind-staging
$ flux check --pre
► checking prerequisites
✔ kubectl 1.21.0 >=1.18.0-0
✔ Kubernetes 1.18.2 >=1.16.0-0
✔ prerequisites checks passed
$ 
$ flux bootstrap github \
>     --token-auth \
>     --context=kind-staging \
>     --owner=${GITHUB_USER} \
>     --repository=${GITHUB_REPO} \
>     --branch=main \
>     --personal \
>     --path=clusters/staging \
>     --verbose
► connecting to github.com
✗ failed to create new Git repository "https://github.com/MxNxPx/flux2-kustomize-helm-example.git": multiple errors occurred: 
- POST https://api.github.com/user/repos: 422 Repository creation failed. [{Resource:Repository Field:name Code:custom Message:name already exists on this account}]
- resource already exists, cannot create object. Use Reconcile() to create it idempotently
```

**Comment by kingdonb** [comment, marked as answer]
I got this message myself yesterday or the day before, I was able to work past it by running `flux uninstall` before bootstrapping again.

I am not sure if this is a bug. Make sure that any Flux CRDs on the cluster are cleaned up before bootstrapping. You may encounter this situation if the cluster has the Flux-system repo defined in it when bootstrap happens.

**Reply by MxNxPx** [reply]
in my case it is a net new kind cluster.  nothing deployed to it at all when i am getting this error.  @kingdonb - do you have the exact error you were getting?  perhaps the error is misleading that i am getting?

again, this works totally fine on v0.11.0 but not on any v0.12.x versions of the flux cli.

**Reply by kingdonb** [reply]
No, unfortunately, I experienced this error when I was working on my Kubecon talk and I had to get it ready to push out by 5pm deadline yesterday, I wasn't doing the usual appropriate follow-up on things that came up.

```
422 Repository creation failed. [{Resource:Repository Field:name Code:custom Message:name already exists on this account}]
- resource already exists, cannot create object. Use Reconcile() to create it idempotently
```

It was definitely this error, I remember reading it and being surprised when it was apparently fixed by running flux uninstall.

**Reply by hiddeco** [reply]
Can you both share the permission scopes of the PAT you were using? As we only attempt to create a repository if it was not found: https://github.com/fluxcd/flux2/blob/main/internal/bootstrap/bootstrap_provider.go#L352-L365

**Reply by kingdonb** [reply]
I've been having trouble narrowing down the PAT, it looks like I'm using the token that my GH cli has created for me. It isn't listed in the PAT section of my account (and it needs to be updated, as it doesn't contain one of the new token-type tags)

The information on the most recent Read-Only deploy key I can find that would have been created at the time by the same GH token, is 
> Added on Apr 19, 2021 by GitHub CLI with authorization from @kingdonb

It is possible that when I got this error, the repo was already created but without any pushed branches. I'm not sure what else could have resulted in this error, (the error that I am sure I was able to recover from without deleting the repo and without generating a new token.)

I'm going to clear all of these tokens today and start again, I will report back if I'm able to trigger the issue once again. Thanks for the helpful reference from the code.

**Reply by MxNxPx** [reply]
Here are the details of my PAT scopes set...

Note: flux

Select scopes:
 repo Full control of private repositories
 repo:status Access commit status
 repo_deployment Access deployment status
 public_repo Access public repositories
 repo:invite Access repository invitations
 security_events Read and write security events
 
 Although in my case i do NOT want it to create a new repo, I want it to use an existing one that I specified but it fails with that error about a new repo...

Thanks!
-mp

**Reply by stefanprodan** [reply]
For existing repos you could use `flux bootstrap git` that doesn't make any API calls to GitHub, see https://toolkit.fluxcd.io/guides/installation/#generic-git-server

**Reply by MxNxPx** [reply]
thanks @stefanprodan!

for using an existing personal github repo, this worked for me on the newer flux cli versions --
```
flux bootstrap git \
    --token-auth=true \
    --context=kind-staging \
    --url=https://github.com/${GITHUB_USER}/${GITHUB_REPO} \
    --username=${GITHUB_USER} \
    --password=${GITHUB_PWD} \
    --branch=main \
    --path=clusters/staging \
    --verbose
```

**Reply by pcornelissen** [reply]
flux version 0.32.0 still has that problem. I get the exact same error message on my local docker for mac with kubernetes enabled.
(Completele new installation)

I also removed the newly created repo, always the same result:


```
flux bootstrap github --owner=orchit --repository=MyRepo --path=projects/MyRepo --reconcile
► connecting to github.com
✔ repository "https://github.com/orchit/MyRepo" created
✗ failed to reconcile Git repository "https://github.com/orchit/MyRepo": multiple errors occurred:
- POST https://api.github.com/orgs/orchit/repos: 422 Repository creation failed. [{Resource:Repository Field:name Code:custom Message:name already exists on this account}]
- resource already exists, cannot create object. Use Reconcile() to create it idempotently
```

**Reply by kingdonb** [reply]
@pcornelissen Can you try adding the `--personal` flag and see if this resolves it?

```
      --personal                if true, the owner is assumed to be a GitHub user; otherwise an org
```

I think this implies that the default is not `--personal` – so without this flag, Flux will look for an organization by the name in `--owner` and try to create a repository there with the name `MyRepo` – until it sees this precondition has succeeded.

This seems like a place where we could do somewhat better for the users. "422 Repository creation failed" when the repo exists is a pretty inscrutable way to fail.

FWIW, I tested it myself and I got this error which is a lot clearer about what has gone wrong ~(I wonder why the difference?)~

```
$ flux bootstrap github --owner kingdonb --repository test-test-flux --path=clusters/my-test-cluster
Please enter your GitHub personal access token (PAT):
► connecting to github.com
✗ failed to create new Git repository "https://github.com/kingdonb/test-test-flux": multiple errors occurred:
- POST https://api.github.com/orgs/kingdonb/repos: 404 Not Found []
- the requested resource was not found
 ```

So that's really clear, at least, about what it's doing... I guess when the repo already exists, ??? – I just tested this too, and I got no error.

```
$ flux bootstrap github --owner kingdonb --repository bootstrap-repo --path=clusters/my-test-cluster
Please enter your GitHub personal access token (PAT):
► connecting to github.com
► cloning branch "main" from Git repository "https://github.com/kingdonb/bootstrap-repo.git"
✔ cloned repository
► generating component manifests
✔ generated component manifests
✔ committed sync manifests to "main" ("2cf0b5b03380d222eeabfd8b2df784d211887a65")
► pushing component manifests to "https://github.com/kingdonb/bootstrap-repo.git"
► installing components in "flux-system" namespace
✔ installed components
✔ reconciled components
► determining if source secret "flux-system/flux-system" exists
► generating source secret
✔ public key: ecdsa-sha2-nistp384 AAAAE2VjZHNhLXNoYTItbmlzdHAzODQAAAAIbmlzdHAzODQAAABhBB3GtzvF9mmHbbv5Np07mcAascth4n6huSzaVl/ErIAqFSiVhOUGL59lh8mCS5wLsUmJ2UbkFD3V7r4x8TGkRUAQb5Jw2071r+dyB8m8h7m2SxOPeD0GrdisjQfaNh6PQQ==
✔ configured deploy key "flux-system-main-flux-system-./clusters/my-test-cluster" for "https://github.com/kingdonb/bootstrap-repo"
► applying source secret "flux-system/flux-system"
✔ reconciled source secret
► generating sync manifests
✔ generated sync manifests
✔ committed sync manifests to "main" ("84a97760623debff84191cab7deff43632e9043d")
► pushing sync manifests to "https://github.com/kingdonb/bootstrap-repo.git"
► applying sync manifests
✔ reconciled sync configuration
◎ waiting for Kustomization "flux-system/flux-system" to be reconciled
✔ Kustomization reconciled successfully
► confirming components are healthy
✔ helm-controller: deployment ready
✔ kustomize-controller: deployment ready
✔ notification-controller: deployment ready
✔ source-controller: deployment ready
✔ all components are healthy
```

Can you explain @pcornelissen what the 422 error is from? (Maybe it needs `--personal=false`?) – is this a repository in a personal repo or in an org?

**Reply by pcornelissen** [reply]
I am sorry, the problem seems to be a missing right in the PAT. I have created a new one with more rights and then it worked

**Reply by kingdonb** [reply]
Superb, to close this thread I'd better call back to this information which sums up:

> we only attempt to create a repository if it was not found

422 error means authorization failure, so a PAT missing the create repo credentials will get that error, and it seems it tells what went wrong. Thanks for closing the loop!
